GoogleDataTransport/GDTCORLibrary/GDTCORFlatFileStorage.m

```diff
@@ -307,24 +323,33 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder {
 
   // Create the singleton and populate its ivars.
   GDTCORStorage *sharedInstance = [self.class sharedInstance];
-  dispatch_sync(sharedInstance.storageQueue, ^{
-    NSSet *classes = [NSSet setWithObjects:[NSMutableOrderedSet class], [GDTCOREvent class], nil];
-    sharedInstance->_storedEvents = [aDecoder decodeObjectOfClasses:classes
-                                                             forKey:kGDTCORStorageStoredEventsKey];
-    classes = [NSSet
-        setWithObjects:[NSMutableDictionary class], [NSMutableSet class], [GDTCOREvent class], nil];
-    sharedInstance->_targetToEventSet =
-        [aDecoder decodeObjectOfClasses:classes forKey:kGDTCORStorageTargetToEventSetKey];
-    sharedInstance->_uploadCoordinator =
-        [aDecoder decodeObjectOfClass:[GDTCORUploadCoordinator class]
-                               forKey:kGDTCORStorageUploadCoordinatorKey];
-  });
+  NSSet *classes = [NSSet setWithObjects:[NSMutableOrderedSet class], [NSMutableDictionary class],
+                                         [GDTCOREvent class], nil];
+  id storedEvents = [aDecoder decodeObjectOfClasses:classes forKey:kGDTCORStorageStoredEventsKey];
+  NSMutableDictionary<NSNumber *, GDTCOREvent *> *events = [[NSMutableDictionary alloc] init];
+  if ([storedEvents isKindOfClass:[NSMutableOrderedSet class]]) {
+    [(NSMutableOrderedSet *)storedEvents
+        enumerateObjectsUsingBlock:^(GDTCOREvent *_Nonnull obj, NSUInteger idx,
+                                     BOOL *_Nonnull stop) {
+          events[obj.eventID] = obj;
+        }];
+  } else if ([storedEvents isKindOfClass:[NSMutableDictionary class]]) {
+    events = (NSMutableDictionary *)storedEvents;
+  }
+  sharedInstance->_storedEvents = events;
```

Comment:
Does it still have to be run on `storageQueue`?

Reply:
No, since appWillForeground was put onto the queue, this will cause a deadlock. appWillForeground is really the more appropriate place anyway.